rs/sns/cli/src/register_extension.rs

```diff
@@ -0,0 +1,521 @@
+        args.map(|args| encode_upgrade_args_without_service(args).unwrap())
+    };
+
+    std::io::stdout().flush().unwrap();
+    std::io::stderr().flush().unwrap();
+
+    Ok(canister_arg)
+}
+
+/// Attempts to create an empty canister on the same subnet as `next_to`.
+///
+/// Returns the ID of the newly created canister in the Ok result.
+pub async fn create_extension_canister<C: CallCanisters>(
+    agent: &C,
+    fiduciary_subnet_id: Option<PrincipalId>,
+    controllers: Vec<PrincipalId>,
+    cycles_amount: u128,
+    name: &str,
+) -> Result<CanisterId> {
+    let subnet = if let Some(subnet) = fiduciary_subnet_id {
+        subnet
+    } else {
+        PrincipalId::from_str("pzp6e-ekpqk-3c5x7-2h6so-njoeq-mt45d-h3h6c-q3mxf-vpeq5-fk5o7-yae")
+            .unwrap()
+    };
+
+    let subnet = SubnetId::new(subnet);
+
+    let subnet_selection = Some(SubnetSelection::Subnet { subnet });
+
+    let canister_id = cycles_ledger_create_canister(
+        agent,
+        cycles_amount,
+        subnet_selection,
+        Some(CanisterSettingsArgs {
+            controllers: Some(BoundedVec::new(controllers)),
+            ..Default::default()
+        }),
+    )
+    .await
+    .map_err(|err| {
+        if let CreateCanisterError::InsufficientFunds { balance } = err {
+            let err = format!(
+                "Requested creating the {} canister with {} cycles, but the caller identity has \
+                 only {} cycles on the cycles ledger. Please buy more cycles using \
+                 `dfx cycles convert --amount AMOUNT --network NETWORK` and try again.",
+                name, cycles_amount, balance,
+            );
+            anyhow::anyhow!(err)
+        } else {
+            anyhow::anyhow!(format!("{:?}", err))
+        }
+    })?
+    .canister_id;
+
+    CanisterId::try_from_principal_id(canister_id).map_err(|err| anyhow::anyhow!(err))
+}
+
+#[derive(Debug, Error)]
+pub enum UpgradeSnsControlledCanisterError<AgentError> {
+    #[error("agent interaction failed: {0}")]
+    Agent(AgentError),
+    #[error("observed bad state: {0}")]
+    Client(String),
+}
+
+impl<E: std::error::Error> From<E> for UpgradeSnsControlledCanisterError<E> {
+    fn from(err: E) -> Self {
+        Self::Agent(err)
+    }
+}
+
+pub struct RegisterExtensionInfo {
+    pub wasm_module_hash: Vec<u8>,
+    pub proposal_id: Option<ProposalId>,
+    pub extension_canister_id: CanisterId,
+}
+
+pub async fn find_sns<C: CallCanisters>(
+    agent: &C,
+    sns_root_canister_id: CanisterId,
+) -> Result<Option<Sns>, UpgradeSnsControlledCanisterError<C::Error>> {
+    let canister_id = sns_root_canister_id.get();
+
+    let root_canister = sns::root::RootCanister { canister_id };
+
+    let response = root_canister.list_sns_canisters(agent).await?;
+    let SnsCanisters { sns, dapps } =
+        SnsCanisters::try_from(response).map_err(UpgradeSnsControlledCanisterError::Client)?;
+
+    Ok(Some(sns))
+}
+
+pub async fn exec<C: CallCanisters>(
```

Comment:
is this the method executing the proposal? would it be good if it has a name that says in a bit more detail what it does?